src/unix/linux/x11/audio.c

```diff
@@ -26,17 +27,18 @@ struct MTY_Audio {
 	size_t pos;
 };
 
-MTY_Audio *MTY_AudioCreate(uint32_t sampleRate, uint32_t minBuffer, uint32_t maxBuffer, uint8_t channels,
-	const char *deviceID, bool fallback)
+MTY_Audio *MTY_AudioCreate(const MTY_AudioFormat *format, uint32_t minBuffer,
+	uint32_t maxBuffer, const char *deviceID, bool fallback)
 {
 	if (!libasound_global_init())
 		return NULL;
 
 	MTY_Audio *ctx = MTY_Alloc(1, sizeof(MTY_Audio));
-	ctx->sample_rate = sampleRate;
-	ctx->channels = channels;
+	ctx->sample_format = format->sampleFormat;
+	ctx->sample_rate = format->sampleRate;
+	ctx->channels = format->channels;
 
-	uint32_t frames_per_ms = lrint((float) sampleRate / 1000.0f);
+	uint32_t frames_per_ms = lrint((float) format->sampleRate / 1000.0f);
```

Comment:
Needs to be multiplied by channes.

Reply:
After starting at the code for an extra bit, it seems to me that I _should not_ multiply by channels. Here's why:

For apple and X11, the `min_buffer` and `max_buffer` values are interpreted as "minimum frame count" and "maximum frame count" respectively. That is, they are describing buffer capacity in units and not in bytes.

On the other hand, our android code is working in bytes and not frame count.

Compare the X11 and android versions, here's an excerpt from `MTY_AudioQueue`:
```c
// X11
	size_t size = count * ctx->frame_size;

	uint32_t queued = audio_get_queued_frames(ctx);

	// Stop playing and flush if we've exceeded the maximum buffer or underrun
	if (ctx->playing && (queued > ctx->max_buffer || queued == 0))
		MTY_AudioReset(ctx);

// Android
	size_t data_size = count * ctx->frame_size;

	audio_start(ctx);

	MTY_MutexLock(ctx->mutex);

	if (ctx->size + data_size >= ctx->max_buffer)
		ctx->flushing = true;
```

Note how in the X11 case, the way we test for exceeding maximum buffer is by calculating the _number_ of queued frames whereas in the Android case, we are doing a comparison with bytes, which takes into account channel count.

Let me know if you disagree and why. I concede that I could be missing something.

Reply:
This should be tested properly, because if we go to 64 channels for example, if it's not correctly multiplied, it will be an issue way faster, with 2 channels, it would be a lesser problem, with more channels it will cause more issues, faster.

If it's samples required here, it should be renamed. Naming in audio stuff is important.

Reply:
Renaming makes sense, I've now done it.

I have tested this thoroughly for Windows and macOS and everything is working correctly. The frame queueing logic for linux is almost exactly the same, and it should work fine. In any case, linux only can do 2 channels anyway. Android I can't test myself.